### PR TITLE
Autocomplete for @mention 

### PR DESCRIPTION
**Pull request by omi10859**
issue: https://github.com/zulip/zulip-terminal/issues/222

### CODE REVIEW

**Comment by neiljp** [comment]
@omi10859 Is this ready for a review?

**Comment by omi10859** [comment]
Not yet. Some work is left will try to finish before coming Monday

**Comment by omi10859** [comment]
@neiljp It's ready for reviews.

**Comment by omi10859** [comment]
LIttle demo:
ezgif com-video-to-gif
For now, pressing `tab` will change to the available options. 
Suggestion/Improvements are welcome.

File: zulipterminal/ui_tools/boxes.py [hunk 1]
```diff
@@ -95,26 +122,28 @@ def keypress(self, size: Tuple[int, int], key: str) -> str:
...
+        #         return key
+        #     # toggle focus position
+        #     if self.focus_position == 0 and self.to_write_box is None:
+        #         if self.contents[0][0].focus_col == 0:
+        #             self.contents[0][0].focus_col = 1
+        #             return key
+        #         else:
+        #             self.contents[0][0].focus_col = 0
+        #     self.focus_position = self.focus_position == 0
+        #     self.contents[0][0].focus_col = 0\
```
[the remark this answers is not in the document]
Comment (omi10859) [reply zulipterminal/ui_tools/boxes.py]: >  I assume this is still here since you're not sure what to do about the key conflict?

Yeah

**Comment by omi10859** [comment]
>  So, we need to determine another good key to use to make this able to be merged.

I was thinking to add a kind of scrollable pop-up with all the possible options similar to what we have in zulip-desktop. Will have to look at technical expects of this, will have to do little research. 

ps: I'm little busy with my exams so this may take little time

**Comment by omi10859** [comment]
> gitignore: ignore .DS_Store

Should I make a new PR for this commit? as this doesn't belong here @neiljp

**Comment by neiljp** [comment]
Don't worry about a separate PR for the `.gitignore` commit; we can always cherry-pick this off to merge separately/first, or just merge together.

For now I would focus on tidying what you have - it's already a lot easier than remembering and typing everyone's full name, and a clean working commit would be good to have, to build upon. Perhaps set the completion key to `ctrl j` for now, since that seems to be a completion combination used in editors; we can at least test it in that case, and it would preserve the key we already have for moving between compose boxes.

Once you have that, you could explore the capitalization in an extra commit.

File: zulipterminal/ui_tools/boxes.py [hunk 2]
```diff
@@ -80,6 +86,28 @@ def stream_box_view(self, button: Any=None, caption: str='',
...
 
+    def autocomplete(self, text: Any, state: int) -> Any:
+        if text.startswith('@'):
+            users = self.view.users.copy()
+            users_list = users.copy()
+            for user in users:
+                if not match_user(user, text[1:]):
+                    users_list.remove(user)
+            tmp = []
+            if text:
```
Comment (neiljp) [inline zulipterminal/ui_tools/boxes.py]: Can this condition ever be `False`?

File: zulipterminal/ui_tools/boxes.py [hunk 3]
```diff
@@ -111,7 +139,6 @@ def keypress(self, size: Tuple[int, int], key: str) -> str:
                     self.contents[0][0].focus_col = 0
             self.focus_position = self.focus_position == 0
             self.contents[0][0].focus_col = 0
-
```
Comment (neiljp) [inline zulipterminal/ui_tools/boxes.py]: This line removal seems separate to this commit?

File: zulipterminal/ui_tools/boxes.py [hunk 4]
```diff
@@ -80,6 +86,28 @@ def stream_box_view(self, button: Any=None, caption: str='',
         ]
         self.contents = write_box
 
+    def autocomplete(self, text: Any, state: int) -> Any:
+        if text.startswith('@'):
+            users = self.view.users.copy()
+            users_list = users.copy()
+            for user in users:
+                if not match_user(user, text[1:]):
+                    users_list.remove(user)
```
Comment (neiljp) [inline zulipterminal/ui_tools/boxes.py]: Can you combine this with the loop below? It seems confusing to have two lists here - was there an alternative implementation you were looking at that used this, or a future version which would?

**Comment by neiljp** [comment]
@omi10859 Will you still be working on this?

**Comment by omi10859** [comment]
Yeah, will start working in few days. 
Busy with exams. I hope that won't be a problem.

**Comment by neiljp** [comment]
@omi10859 Thanks for responding; hope the exams go well!

File: zulipterminal/ui_tools/boxes.py (hunk 2 above)
Comment (omi10859) [reply zulipterminal/ui_tools/boxes.py]: this condition will remove all the not matching names from the list of users

File: zulipterminal/ui_tools/boxes.py (hunk 4 above)
Comment (omi10859) [reply zulipterminal/ui_tools/boxes.py]: the first loop is used to remove all the not-matching users from users_list, second loop creates a new list that is iterated by user for the suggestions.

**Comment by neiljp** [comment]
@omi10859 Let us know when this is ready for review. For now, there seem to be some conflict lines (>>, ==) in your commits, and you might consider splitting at least one of the commits, eg. one can be just an upgrade of urwid_readline?

**Comment by omi10859** [comment]
There are total two commits one is for all the coding and another is for changing the Readme file 
>though it could also be combined with the main commit.

In the first commit, I mistakenly committed conflict lines and removed in next commit. 
Will fix that

**Comment by sumanthvrao** [comment]
@omi10859 You could consider splitting the first commit into smaller once -> The first for upgrading `urwid_readline` version, and other for key addition, etc.

File: README.md [hunk 5]
```diff
@@ -121,6 +121,7 @@ autohide=autohide
 | Jump to next line                                     | <kbd>Ctrl</kbd> + <kbd>N</kbd> / <kbd>↓</kbd> |
 | Quit                                                  | <kbd>Ctrl</kbd> + <kbd>C</kbd>                |
 | Clear screen                                          | <kbd>Ctrl</kbd> + <kbd>L</kbd>                |
+| Auto Complete                                         | <kbd>Shift</kbd> + <kbd>Tab</kbd>             |
```
Comment (shreyamalviya) [inline README.md]: I think the description over here should be a little more elaborate. Autocompletes what?

File: zulipterminal/ui_tools/boxes.py (hunk 4 above)
Comment (shreyamalviya) [reply zulipterminal/ui_tools/boxes.py]: @omi10859, I believe you could do that without two loops too.
```python3
tmp = []
for user in users:
    if match_user(user, text[1:]):
        tmp.append('@**%s**' % user['full_name'])
```

**Comment by neiljp** [comment]
@omi10859 You still have conflict lines added and removed in both commits.

It is a bit difficult to review this fully when you still have conflicts in your commits. Also please extract the changes causing the upgrade of urwid-readline into a separate commit which applies first. If you have troubles with this then please let us know and we can help, or you could discuss on various chat.zulip.org streams (and elsewhere, of course).

File: zulipterminal/ui_tools/boxes.py (hunk 4 above)
Comment (omi10859) [reply zulipterminal/ui_tools/boxes.py]: Can you suggest me a way?

File: README.md (hunk 5 above)
Comment (omi10859) [reply README.md]: Autocomplete will be used for autocompleting streams and mentions. Can you suggest a better description?

**Comment by sumanthvrao** [comment]
@omi10859 Are you continuing work on this?

**Comment by neiljp** [comment]
@omi10859 Your work from this PR has been included as a commit as part of #441, so I'm closing this PR now. If you intend to work further on this area please let us know on chat.zulip.org as this is now being actively developed further based on your initial work :)

**Pull request closed without merging**